// @ohos/hvigor-ohos-plugin/index.js
const path = require('path');
const fs = require('fs');

const toolsRoot = path.resolve(
  process.env.DEVECO_TOOLS_ROOT ||
  process.env.DEVECO_TOOLS_HOME ||
  'C:/Program Files/Huawei/DevEco Studio/tools'
);
const sdkRoot = path.resolve(
  process.env.TABSSH_HWSDK_ROOT ||
  process.env.HARMONYOS_SDK_ROOT ||
  process.env.DEVECO_SDK_HOME ||
  'C:/Program Files/Huawei/DevEco Studio/sdk/default'
);
const pluginRoot = path.resolve(toolsRoot, 'hvigor/hvigor-ohos-plugin');

function apiVersion(value) {
  return {
    getMajor: () => value,
    getValue: () => value,
    equals: (other) => Number(other && (other.getValue ? other.getValue() : other)) === value,
    compareTo: (other) => value - (Number(other && (other.getValue ? other.getValue() : other)) || 0),
    greaterThan: (other) => value > (Number(other && (other.getValue ? other.getValue() : other)) || 0),
    greaterThanOrEquals: (other) => value >= (Number(other && (other.getValue ? other.getValue() : other)) || 0),
    lessThan: (other) => value < (Number(other && (other.getValue ? other.getValue() : other)) || 0),
    lessThanOrEquals: (other) => value <= (Number(other && (other.getValue ? other.getValue() : other)) || 0),
    toString: () => String(value),
    valueOf: () => value
  };
}

function firstExistingDir(candidates) {
  for (const candidate of candidates) {
    if (fs.existsSync(candidate)) return candidate;
  }
  return candidates[0];
}

function patchSdkInfoExport(value) {
  if (typeof value !== 'function' || !value.prototype) return;
  const proto = value.prototype;
  const openharmonyRoot = path.resolve(sdkRoot, 'openharmony');
  const hmsRoot = path.resolve(sdkRoot, 'hms');
  const nativeDir = firstExistingDir([
    path.resolve(openharmonyRoot, 'native'),
    path.resolve(hmsRoot, 'native')
  ]);
  const toolchainsDir = firstExistingDir([
    path.resolve(openharmonyRoot, 'toolchains'),
    path.resolve(hmsRoot, 'toolchains')
  ]);
  const etsDir = firstExistingDir([
    path.resolve(openharmonyRoot, 'ets'),
    path.resolve(hmsRoot, 'ets')
  ]);

  proto.contains = function () { return true; };
  proto.getBaseDir = function () { return openharmonyRoot; };
  proto.getSdkRoot = function () { return sdkRoot; };
  proto.getSdkDir = function () { return openharmonyRoot; };
  proto.getSdkNativeDir = function () { return nativeDir; };
  proto.getNativeDir = function () { return nativeDir; };
  proto.getSdkToolchainsDir = function () { return toolchainsDir; };
  proto.getToolchainsDir = function () { return toolchainsDir; };
  proto.getSdkEtsDir = function () { return etsDir; };
  proto.getEtsDir = function () { return etsDir; };
  proto.getApiVersion = function () { return apiVersion(24); };
  proto.getFullApiVersion = function () { return apiVersion(24); };
}

function patchSdkInfoModule(relativePath) {
  try {
    const loaded = require(path.resolve(pluginRoot, relativePath));
    patchSdkInfoExport(loaded);
    patchSdkInfoExport(loaded.default);
    patchSdkInfoExport(loaded.OhosSdkInfo);
    patchSdkInfoExport(loaded.HmosSdkInfo);
    patchSdkInfoExport(loaded.HmsSdkInfo);
    patchSdkInfoExport(loaded.HosSdkInfo);
  } catch (_) {}
}

patchSdkInfoModule('src/sdk/info/common-sdk-info.js');
patchSdkInfoModule('src/sdk/info/ohos-sdk-info.js');
patchSdkInfoModule('src/sdk/info/hmos-sdk-info.js');
patchSdkInfoModule('src/sdk/info/hms-sdk-info.js');

module.exports = require(pluginRoot);
